python/util.py

```diff
@@ -0,0 +1,257 @@
+            self[i] = FormattedChar(self[i], **{name: fmt})
+            existing_fmt = ''
+        else:
+            if add:
+                fmt = existing_fmt + fmt
+            setattr(self[i], name, fmt)
+
+    def prepend_format_index(self, i, fmt):
+        'Add a format before the indexed character'
+        self._format_index(i, 'pre_fmt', fmt, add=True)
+
+    def pre_format_index(self, i, fmt):
+        'Replace the format before the indexed character'
+        self._format_index(i, 'pre_fmt', fmt)
+
+    def append_format_index(self, i, fmt):
+        'Add a format after the indexed character'
+        self._format_index(i, 'post_fmt', fmt, add=True)
+
+    def post_format_index(self, i, fmt):
+        'Replace the format after the indexed character'
+        self._format_index(i, 'post_fmt', fmt)
+
+    def get_format(self, i):
+        'Get the latest format string to be applied to the indexed character'
+        for c in self[i::-1]:
+            if hasattr(c, 'pre_fmt'):
+                return c.post_fmt or c.pre_fmt
+        return ''
+
+    def __str__(self):
+        return ''.join(map(str, self))
+
+
+class Time(namedtuple('Time', ('hours', 'minutes', 'seconds', 'precision'))):
+    '''Handle rounding and pretty formatting of time. Precision has two related
+    meanings:
+    1. The number of decimal places to which to render seconds
+    2. Which units to show in the output
+    Because of this strong coupling, you can't ask for a "hour-like" time at a
+    high precision.
+
+    Calculating the exact breakdown into hours, minutes and seconds components
+    is complicated by the fact that the precision to which we want to round the
+    seconds component is dependent on whether the seconds component overflows
+    at a given precision!
+    '''
+    _formats = {
+        0: '{t.hours:.0f}:{t.minutes:02.0f}:{t.seconds:02.0f}',
+        1: '{t.minutes:.0f}:{t.seconds:04.1f}',
+        2: '{t.seconds:.2f}',
+    }
+
+    def __new__(cls, hours, minutes, seconds, target_precision=None):
+        if target_precision is None:
+            precision = cls._get_precision(hours, minutes, seconds)
+        else:
+            precision = target_precision
+        hours, minutes, seconds = cls._round(
+            hours, minutes, seconds, precision)
+        precision = cls._get_precision(hours, minutes, seconds)
+        if target_precision is not None:
+            if precision < target_precision:
+                raise ValueError("Precision too high for large time")
+            else:
+                precision = target_precision
+        return super().__new__(cls, hours, minutes, seconds, precision)
+
+    @classmethod
+    def from_seconds(cls, seconds, precision=None):
+        hours, remainder = divmod(seconds, 3600)
+        minutes, seconds = divmod(remainder, 60)
+        return cls(hours, minutes, seconds, precision)
+
+    @staticmethod
+    def _get_precision(hours, minutes, seconds):
+        if not hours:
```

Comment:
That's highly discouraged from a PEP8 viewpoint - it certainly upsets `flake8` in my editor!